src/Microsoft.IdentityModel.Xml/X509Data.cs

```diff
@@ -120,31 +120,24 @@ public string CRL
             set;
         }
 
-        /// <summary>
-        /// Compares two X509Data objects.
-        /// </summary>
+        /// <inheritdoc/>
         public override bool Equals(object obj)
         {
-            var other = obj as X509Data;
-            if (other == null)
-                return false;
-            else if (!IssuerSerial.Equals(other.IssuerSerial) ||
-                string.Compare(SKI, other.SKI, StringComparison.OrdinalIgnoreCase) != 0 || 
-                string.Compare(SubjectName, other.SubjectName, StringComparison.OrdinalIgnoreCase) != 0 ||
-                // certificates may need to be compared in a special way instead of generic string comparison?
-                !Enumerable.SequenceEqual(Certificates.OrderBy(t => t), other.Certificates.OrderBy(t => t)) ||
-                string.Compare(CRL, other.CRL, StringComparison.OrdinalIgnoreCase) != 0)
-                    return false;
-            return true;
+            return obj is X509Data data &&
+                EqualityComparer<IssuerSerial>.Default.Equals(IssuerSerial, data.IssuerSerial) &&
+                string.Equals(SKI, data.SKI, StringComparison.OrdinalIgnoreCase) &&
+                string.Equals(SubjectName, data.SubjectName, StringComparison.OrdinalIgnoreCase) &&
+                Enumerable.SequenceEqual(Certificates.OrderBy(t => t), data.Certificates.OrderBy(t => t)) &&
+                string.Equals(CRL, data.CRL, StringComparison.OrdinalIgnoreCase);
         }
-
-        /// <summary>
-        /// Serves as a hash function for X509Data.
-        /// </summary>
+        /// <inheritdoc/>
         public override int GetHashCode()
         {
-            return base.GetHashCode();
+            unchecked
+            {
+                // Certificates is the only immutable property
```

Comment:
Not sure why this method was originally there, but also not sure why we decided to change it. Was this flagged too? What are the benefits of the new implementation?

Reply:
Sonar has a good write up on the finding

https://rules.sonarsource.com/csharp/type/Bug/RSPEC-3249

Here is the finding:

https://sonarcloud.io/project/issues?id=AzureAD_azure-activedirectory-identitymodel-extensions-for-dotnet&open=AXNs_zk7UUMedrjWA3JI&resolved=false&types=BUG

Reply:
Thanks. I'm still a bit confused about this.
_EqualityComparer<ICollection<string>>.Default.GetHashCode(Certificates)_ returns a different hash code for each X509Data object, right? This implementations satisfies the guideline that GetHashCode function should return the same hash code over the lifetime of the object (immutability), but the following is not true: _Two objects that are equal return hash codes that are equal._

If both arguments above are true, it seems like the is no perfect solution and that we are choosing to be back-compat.

Reply:
GetHashCode with classes that contain collections that are not set by reference in the ctor end up with this limitation. From the GetHashCode perspective, the object is seen as not equal. 

That is my understanding at least.

Reply:
You do raise a good point. There is a slight behavioral change in collections that use the hash code

Reply:
This is interesting as the recommendation is that if obj1.HashCode() ! obj2.HashCode() => obj1 != obj2. But this will not be true, right?

Reply:
@brentschmaltz Correct. I don't think we can get there in this case though. Functionally this is somewhat similar to the previous behavior. Every X509Data object will have a different hashcode